### DISCUSSION TITLE
someone knows of a Salvo tutorial?

### DISCUSSION THREAD

**Discussion by jhjacobs81** [salvo-rs/salvo#174] [Q&A]
I am looking at all the examples, but mixing and matching them seems to give me more problems then i can handle.

Does someone know of a tutorial that uses Salvo to write a CRUD web app? What i am looking for is a basic cms-like idea that uses sessions, user logins with a database backend. I am trying to learn Rust, but its not going very fast, and i need something workable rather sooner then later.

I'm not very rich but i am willing to pay a bit if no such tutorial exists but someone is willing to make such a webapp. preferably with some comments in the code so i can then use it to build upon.

**Comment by chrislearn** [comment]
There are two blog system build with salvo:
https://github.com/driftluo/myblog
https://github.com/prabirshrestha/rblog

**Reply by jhjacobs81** [reply]
they look great, but rather confusing for me :(

SOO.. i'll just start from the bottom up :-)

I have templates working! now i am trying to get basic auth working, but this fails. 
Following this example: https://github.com/salvo-rs/salvo/tree/main/examples/basic-auth

i have in my routes function:
```
struct Validator;
#[async_trait]
impl BasicAuthValidator for Validator {
    async fn validate(&self, username: &str, password: &str) -> bool {
        username == "root" && password == "pwd"
    }
}

// ROUTES
pub fn routes() -> Router {
    let auth_handler = BasicAuth::new(Validator);
    Router::new()
        .get(login_page)
        .push(Router::with_path("app/borg/logs")
        .hoop(auth_handler)
        .get(borg_logs)
    )
}
```

But this leads to an error:
```
rror[E0632]: cannot provide explicit generic arguments when `impl Trait` is used in argument position
  --> /home/sysop/.cargo/registry/src/github.com-1ecc6299db9ec823/salvo_extra-0.37.1/src/force_https.rs:54:42
   |
54 |         if let Some(host) = req.header::<String>(header::HOST) {
   |                                          ^^^^^^ explicit generic argument not allowed
   |
   = note: see issue #83701 <https://github.com/rust-lang/rust/issues/83701> for more information

For more information about this error, try `rustc --explain E0632`.
error: could not compile `salvo_extra` due to previous error
```

**Comment by chrislearn** [comment]
Make sure you are using rust version >= 1.64

**Comment by jhjacobs81** [comment]
boy.. do i feel stupid.. This was indeed the problem! basicauth works flawless now!
next stop is sessions :) then databases.

**Comment by jhjacobs81** [comment]
unfortunatly, i seem to be unable to get the sqlx example to work.
when i use this example: https://pastebin.com/0D6Cm6un
it prints out a list of pages on the commandline.

but when i use this example: https://pastebin.com/v9DmxnwA (which is an almost exact replica of the example) it returns an EMTY_RESPONSE, and the commandline shows this:
```
thread 'tokio-runtime-worker' panicked at 'called `Option::unwrap()` on a `None` value', src/main.rs:23:37
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

It seems that the req.query return None, regardless of what i try :( 

This works!
```
let slug = req.params().get("slug").cloned().unwrap_or_default();
```

**Reply by chrislearn** [reply]
![image](https://user-images.githubusercontent.com/5874864/197641645-c064f3de-d5c1-4205-97ed-fd5b9aa0f3ed.png)

**Reply by jhjacobs81** [reply]
Is there a difference between req.param and req.params, and why then is req.param better? :-) I’m still learning, so very curious :)

**Reply by chrislearn** [reply]
![image](https://user-images.githubusercontent.com/5874864/197666948-cbdc7868-c318-4261-9db8-134fd0ac2ad4.png)
![image](https://user-images.githubusercontent.com/5874864/197666983-fb07f922-2ba0-4370-9ec2-82a4c137f55f.png)
`params` is a HashMap<String, String>, but we may want get a i64 value or any other data type, you can use `req.param::<i64>("id")`, it is  easier than use `req.params().get("id").parse::<i64>().unwrap()`.

**Comment by jhjacobs81** [comment]
this is much true :)

so: params when multiple values, param when single value of wich you know the type, thats basicly it.

**Comment by jhjacobs81** [comment]
so now, i am trying to serve static content following: https://salvo.rs/book/middlewares/serve_static/

i added everything on this page, and the error i receive is:
```
error[E0432]: unresolved import `salvo_extra::serve`
 --> src/routes.rs:4:18
  |
4 | use salvo_extra::serve::StaticDir;
  |                  ^^^^^ could not find `serve` in `salvo_extra`
```

**Reply by chrislearn** [reply]
I'll update this document. You can check this example:

https://github.com/salvo-rs/salvo/blob/main/examples/static-dir-list/src/main.rs

**Comment by jhjacobs81** [comment]
thank you :)

i now have this:
```
        .push(
            Router::with_path("/static/").get(
                StaticDir::new([
                    "static",
                ])
                .with_defaults("index.html")
                .with_listing(true),
            ),
        )
```

where i would think static is the directory under the main project folder (so, next to the src folder) but this throws me an error 404. 
I got logging working though :)

**Reply by chrislearn** [reply]
Try add wildcard match rule like this:

```rust
 .push(
            Router::with_path("/static/<**rest>").get(
                StaticDir::new([
                    "static",
                ])
                .with_defaults("index.html")
                .with_listing(true),
            ),
        )
```

**Reply by jhjacobs81** [reply]
this works :)

**Comment by jhjacobs81** [comment]
Also, i bought you some coffee, did you see? :) 

can you contact me privately? I have some questions, and some money. But i rather not discuss it publicly yet!

**Reply by chrislearn** [reply]
I got your coffee, thank you very much. You can contact my email: chris@acroidea.com or Telegram: chrislearn_young

**Comment by jhjacobs81** [comment]
so i have this:
```rust
#[derive(Debug, Deserialize)]
pub struct LoginUser {
    pub username: String,
    pub password: String,
}
```

and:
```rust
#[handler]
pub async fn login(req: &mut Request, depot: &mut Depot, res: &mut Response) {
    if req.method() == salvo::http::Method::POST {
        // Get the username and password from the database
        // If they are correct, get the user role and push it into the session for use everywhere. 
        // This will allow users to access information based on their role
        let uname = req.form::<String>("username").await;
        let upass = req.form::<String>("password").await;
        let user = sqlx::query_as::<_, LoginUser>("select * from users where username = $1 AND password = $2");
    

        }
    else {
        let loginpage = LoginTemplate {
            text: "Please login!",
         };
        res.render(Text::Html(loginpage.render().unwrap()));
    }
}
```

but it wont compile because i get the error:
```rust
21  |         let user = sqlx::query_as::<_, LoginUser>("select * from users where username = $1 AND password = $2");
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `for<'r> FromRow<'r, _>` is not implemented for `structs::general::LoginUser` 
```

**Reply by chrislearn** [reply]
This error is related to sqlx not salvo, You should implement FromRow for LoginUser.
See this example: https://github.com/salvo-rs/salvo/blob/main/examples/postgres-sqlx/src/main.rs

```rust
#[derive(FromRow, Serialize, Debug)]
pub struct User {
    pub id: i64,
    pub username: String,
    pub password: String,
}
```

Notice FromRow in derive.

**Comment by jhjacobs81** [comment]
It's true that this is not a Salvo problem, however i already added FromRow, which gives out different errors :) Maybe i'm just not cut out to learn to program :)

**Comment by jhjacobs81** [comment]
ok, so now i have a different problem :-)

code:
```rust // ADD DATA TO DATABASE
#[handler]
pub async fn borg_postlogs(res: &mut Response, req: &mut Request) {
    // POST incoming data to the database
    #[derive(Deserialize, Debug)]
    struct Data {
    token: String,
    date: String,
    name: String,
    hostname: String,
    duration: String,
    files: String,
    compressed_size: String,
    original_size: String,
    repository_id: String,
    backupdir: String,
    }

    let d = req.parse_json::<Data>();
    let e = d.token;
    res.render(Text::Html(e));
}
```

error: 
```rust sysop@rust-deb:~/rust-prod/UNIMATRIX$ cargo run
   Compiling unimatrix v0.1.0 (/home/sysop/rust-prod/UNIMATRIX)
error[E0609]: no field `token` on type `impl Future<Output = Result<borg::borg_postlogs::borg_postlogs::{closure#0}::Data, salvo::http::ParseError>>`
  --> src/handlers/borg.rs:73:15
   |
73 |     let e = d.token;
   |               ^^^^^ field not found in `impl Future<Output = Result<borg::borg_postlogs::borg_postlogs::{closure#0}::Data, salvo::http::ParseError>>`

For more information about this error, try `rustc --explain E0609`.
error: could not compile `unimatrix` due to previous error
```

if i'm to believe "rustc --explain E0609" everything should work?

**Reply by chrislearn** [reply]
try this:

```rust
#[handler]
pub async fn borg_postlogs(res: &mut Response, req: &mut Request) {
...
  }

  let d = req.parse_json::<Data>().await;
  let e = d.token;
  res.render(Text::Html(e));
}
```

**Reply by chrislearn** [reply]
You need to add await after call `parse_json`.

**Comment by jhjacobs81** [comment]
oh thats what i had originally, but that doesnt help either.  What i forgot was the .unwrap thingy :|
I spend a few hours looking at that..
